File: src/js/lib/bits.js
// Copyright (c) Tim Gordon.
// This file is licensed to you under the Apache Licence, Version 2.0. See the LICENSE file.

/**
 * Bit primitives over a 32 bit word.
 *
 * The C# implementation works in 64 bit words because .NET has an integer that wide. JavaScript
 * does not: its bitwise operators convert to a signed 32 bit integer first, so the natural word
 * here is half the size and there are twice as many of them. Using BigInt to keep the widths the
 * same would be slower for no gain, since nothing in a character set needs more than a bit per
 * character.
 *
 * Every function takes whatever a bitwise operator produced, which is a signed 32 bit integer, and
 * treats it as 32 unsigned bits.
 */

/**
 * The number of set bits in a word.
 *
 * @param {number} value The word, read as 32 unsigned bits.
 * @returns {number} The number of set bits, from 0 to 32.
 */
export function popCount(value) {
	// The usual SWAR sum: pairs, then nibbles, then a multiply that sums the bytes into the top
	// byte. `Math.imul` is used for the multiply because an ordinary `*` would produce a value
	// above 2^32 and the shift below would then have to truncate it.
	let bits = value >>> 0;

	bits = bits - ((bits >>> 1) & 0x55555555);
	bits = (bits & 0x33333333) + ((bits >>> 2) & 0x33333333);
	bits = (bits + (bits >>> 4)) & 0x0f0f0f0f;

	return Math.imul(bits, 0x01010101) >>> 24;
}

/**
 * The number of zero bits below the least significant set bit of a word, or 32 if the word is
 * zero.
 *
 * @param {number} value The word, read as 32 unsigned bits.
 * @returns {number} The trailing zero count, from 0 to 32.
 */
export function trailingZeroCount(value) {
	const bits = value | 0;

	if (bits === 0) { return 32; }

	// Isolating the lowest set bit leaves a single bit, and one less than that is a run of
	// exactly as many ones as there were trailing zeros. Where that single bit is bit 31 the
	// subtraction leaves the range of a signed 32 bit integer, which `popCount` folds back.
	return popCount((bits & -bits) - 1);
}

File: src/cs/Naxp/Bits.cs
// Copyright (c) Tim Gordon.
// This file is licensed to you under the Apache Licence, Version 2.0. See the LICENSE file.

using System.Runtime.CompilerServices;
#if NET8_0_OR_GREATER
using System.Numerics;
#endif

namespace LogMu;

/// <summary>
/// Bit primitives over <see cref="ulong"/>.
/// <para>
/// On net8.0 these map to <c>System.Numerics.BitOperations</c>, and so to hardware instructions.
/// netstandard2.0 has no <c>BitOperations</c>, so each method carries a software version in its
/// <c>#else</c> branch. Only one of the two is compiled for a given target, so the software
/// version is covered by the net472 test run, which loads the netstandard2.0 build.
/// </para>
/// </summary>
static class Bits
{
	/// <summary>
	/// The number of set bits in <paramref name="value"/>.
	/// </summary>
	/// <param name="value">The value whose set bits are to be counted.</param>
	/// <returns>The number of set bits, in the range 0 to 64.</returns>
	[MethodImpl(MethodImplOptions.AggressiveInlining)]
	public static int PopCount(ulong value)
	{
#if NET8_0_OR_GREATER
		return BitOperations.PopCount(value);
#else
		// The usual SWAR sum: pairs, then nibbles, then a multiply that sums the bytes into the top byte.
		value -= (value >> 1) & 0x5555555555555555UL;
		value = (value & 0x3333333333333333UL) + ((value >> 2) & 0x3333333333333333UL);
		value = (value + (value >> 4)) & 0x0F0F0F0F0F0F0F0FUL;
		return (int)((value * 0x0101010101010101UL) >> 56);
#endif
	}
	/// <summary>
	/// The number of zero bits below the least significant set bit of <paramref name="value"/>,
	/// or 64 if <paramref name="value"/> is zero.
	/// </summary>
	/// <param name="value">The value to examine.</param>
	/// <returns>The trailing zero count, in the range 0 to 64.</returns>
	[MethodImpl(MethodImplOptions.AggressiveInlining)]
	public static int TrailingZeroCount(ulong value)
	{
#if NET8_0_OR_GREATER
		return BitOperations.TrailingZeroCount(value);
#else
		if (value == 0UL) { return 64; }

		// Isolating the lowest set bit leaves a single bit, and one less than that
		// is a run of exactly as many ones as there were trailing zeros.
		return PopCount((value & (0UL - value)) - 1UL);
#endif
	}
}
